File: backend/src/schemas/sewing.py
"""Pydantic shapes for the Sewing (remessas) feature.

A sewing shipment is a bundle of cut pieces sent from the factory to a
banca (contractor). The banca delivers (possibly partially) and the
received pieces become stock entries automatically.
"""

import uuid
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator

from models.enums import ShipmentStatus, Size
from schemas._common import Page


class ShipmentItemInput(BaseModel):
    """One per-size line in the create payload."""

    size: Size
    requested_quantity: int = Field(ge=0)


class ShipmentItemReceiveInput(BaseModel):
    """One per-size line in the receive payload."""

    size: Size
    received_quantity: int = Field(ge=0)


class ShipmentItemRead(BaseModel):
    id: uuid.UUID
    size: Size
    requested_quantity: int
    received_quantity: int


class ShipmentCuttingOrderRead(BaseModel):
    """Inline cutting-order metadata embedded in ShipmentRead."""

    id: uuid.UUID
    code: str


class ShipmentContractorRead(BaseModel):
    """Inline contractor metadata embedded in ShipmentRead."""

    id: uuid.UUID
    name: str


class ShipmentCreate(BaseModel):
    cutting_order_id: uuid.UUID
    contractor_id: uuid.UUID
    sent_at: date
    items: list[ShipmentItemInput] = Field(min_length=1)

    @field_validator("items")
    @classmethod
    def _items_unique_sizes_and_positive(cls, items: list[ShipmentItemInput]) -> list[ShipmentItemInput]:
        sizes = [item.size for item in items]
        if len(set(sizes)) != len(sizes):
            raise ValueError("Each size may appear only once in items")
        if not any(item.requested_quantity > 0 for item in items):
            raise ValueError("At least one item must have a quantity greater than zero")
        return items


class ShipmentReceiveBody(BaseModel):
    received_at: date
    items: list[ShipmentItemReceiveInput] = Field(min_length=1)

    @field_validator("items")
    @classmethod
    def _items_unique_sizes(
        cls,
        items: list[ShipmentItemReceiveInput],
    ) -> list[ShipmentItemReceiveInput]:
        sizes = [item.size for item in items]
        if len(set(sizes)) != len(sizes):
            raise ValueError("Each size may appear only once in receive items")
        return items


class ShipmentRead(BaseModel):
    """Eager-loaded shipment representation for the API."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    cutting_order: ShipmentCuttingOrderRead
    contractor: ShipmentContractorRead
    status: ShipmentStatus
    sent_at: date
    received_at: date | None
    items: list[ShipmentItemRead]
    created_at: datetime
    updated_at: datetime


class ShipmentFilters(BaseModel):
    q: str | None = Field(default=None, max_length=120)
    status: ShipmentStatus | None = None
    contractor_id: uuid.UUID | None = None
    cutting_order_id: uuid.UUID | None = None


ShipmentPage = Page[ShipmentRead]

File: frontend/src/lib/schemas/sewing.ts
/**
 * Zod schemas for the Sewing (Costura) feature.
 *
 * Mirrors `backend/src/schemas/sewing.py`. A Shipment (remessa) bundles cut
 * pieces sent from a DONE cutting order to a contractor (banca); the wire shape
 * embeds compact cutting-order + contractor projections so list rows can render
 * in one fetch.
 *
 * Receiving is now PARTIAL + repeatable: each item carries `requested`,
 * `received`, and `credited` (the watermark of how much of `received` has
 * already been posted to blank-piece stock). A receive only credits the new
 * delta (`received − credited`); re-calling tops up.
 */

import { z } from "zod";
import { SIZES, type Size } from "@/lib/schemas/product";

export const SHIPMENT_STATUSES = [
  "sent",
  "received",
  "partial",
  "cancelled",
] as const;

export type ShipmentStatus = (typeof SHIPMENT_STATUSES)[number];
export const shipmentStatusSchema = z.enum(SHIPMENT_STATUSES);

export const shipmentItemReadSchema = z.object({
  id: z.string(),
  size: z.enum(SIZES),
  requested_quantity: z.number().int().min(0),
  received_quantity: z.number().int().min(0),
  credited_quantity: z.number().int().min(0),
});

export type ShipmentItem = z.infer<typeof shipmentItemReadSchema>;

export const shipmentCuttingOrderRefSchema = z.object({
  id: z.string(),
  code: z.string(),
});

export const shipmentContractorRefSchema = z.object({
  id: z.string(),
  name: z.string(),
});

export const shipmentReadSchema = z.object({
  id: z.string(),
  // Null for legacy standalone remessas whose source cutting order is gone.
  cutting_order: shipmentCuttingOrderRefSchema.nullable(),
  contractor: shipmentContractorRefSchema,
  status: shipmentStatusSchema,
  sent_at: z.string(),
  received_at: z.string().nullable().optional(),
  items: z.array(shipmentItemReadSchema),
  created_at: z.string(),
  updated_at: z.string(),
});

export type Shipment = z.infer<typeof shipmentReadSchema>;

export const shipmentPageSchema = z.object({
  items: z.array(shipmentReadSchema),
  total: z.number(),
  page: z.number(),
  page_size: z.number(),
  has_more: z.boolean(),
});

export type ShipmentPage = z.infer<typeof shipmentPageSchema>;

export type ShipmentFilters = {
  q?: string;
  status?: ShipmentStatus;
  contractor_id?: string;
  cutting_order_id?: string;
  page?: number;
  page_size?: number;
};

/**
 * Form-side schema for creating a shipment. Per-size requested quantities
 * live on a `sizes` map keyed by size; the transform step in
 * `buildShipmentCreatePayload` flattens the map to the backend list shape
 * and drops any size with quantity 0. The cutting order is picked from the
 * available-cuts list; the form clamps each size's `max` to availability as a
 * soft client guard (the backend re-validates against live availability).
 */
const requestedSizeMap = z
  .object({
    p: z.coerce.number().int().min(0).default(0),
    m: z.coerce.number().int().min(0).default(0),
    g: z.coerce.number().int().min(0).default(0),
    gg: z.coerce.number().int().min(0).default(0),
    u: z.coerce.number().int().min(0).default(0),
  })
  .superRefine((data, ctx) => {
    const total = data.p + data.m + data.g + data.gg + data.u;
    if (total <= 0) {
      ctx.addIssue({
        code: "custom",
        path: ["p"],
        message: "validation.atLeastOneSize",
      });
    }
  });

export const shipmentFormSchema = z.object({
  cutting_order_id: z.string().min(1, { message: "validation.cuttingOrderRequired" }),
  contractor_id: z.string().min(1, { message: "validation.contractorRequired" }),
  sent_at: z.string().trim().min(1, { message: "validation.sentAtRequired" }),
  sizes: requestedSizeMap,
});

export type ShipmentFormValues = z.input<typeof shipmentFormSchema>;
export type ShipmentFormParsed = z.output<typeof shipmentFormSchema>;

export type ShipmentCreatePayload = {
  cutting_order_id: string;
  contractor_id: string;
  sent_at: string;
  items: Array<{ size: Size; requested_quantity: number }>;
};

export function buildShipmentCreatePayload(
  parsed: ShipmentFormParsed,
): ShipmentCreatePayload {
  const items: Array<{ size: Size; requested_quantity: number }> = [];
  for (const size of SIZES) {
    const requested_quantity = parsed.sizes[size];
    if (requested_quantity > 0) items.push({ size, requested_quantity });
  }
  return {
    cutting_order_id: parsed.cutting_order_id,
    contractor_id: parsed.contractor_id,
    sent_at: parsed.sent_at,
    items,
  };
}

/**
 * Receive-side schema. Each size carries the *received* count; values must
 * be non-negative and at least one size must have a positive quantity. Sizes
 * omitted from the payload retain their current received count server-side
 * (partial = send fewer/lower sizes, re-call to top up).
 */
export const shipmentReceiveFormSchema = z.object({
  received_at: z.string().trim().min(1, { message: "validation.receivedAtRequired" }),
  sizes: z
    .object({
      p: z.coerce.number().int().min(0).default(0),
      m: z.coerce.number().int().min(0).default(0),
      g: z.coerce.number().int().min(0).default(0),
      gg: z.coerce.number().int().min(0).default(0),
      u: z.coerce.number().int().min(0).default(0),
    })
    .superRefine((data, ctx) => {
      const total = data.p + data.m + data.g + data.gg + data.u;
      if (total <= 0) {
        ctx.addIssue({
          code: "custom",
          path: ["p"],
          message: "validation.atLeastOneSize",
        });
      }
    }),
});

export type ShipmentReceiveFormValues = z.input<typeof shipmentReceiveFormSchema>;
export type ShipmentReceiveFormParsed = z.output<typeof shipmentReceiveFormSchema>;

export type ShipmentReceivePayload = {
  received_at: string;
  items: Array<{ size: Size; received_quantity: number }>;
};

export function buildShipmentReceivePayload(
  parsed: ShipmentReceiveFormParsed,
): ShipmentReceivePayload {
  const items: Array<{ size: Size; received_quantity: number }> = [];
  for (const size of SIZES) {
    const received_quantity = parsed.sizes[size];
    if (received_quantity > 0) items.push({ size, received_quantity });
  }
  return {
    received_at: parsed.received_at,
    items,
  };
}

export function sumRequested(items: ShipmentItem[]): number {
  return items.reduce((acc, it) => acc + it.requested_quantity, 0);
}
export function sumReceived(items: ShipmentItem[]): number {
  return items.reduce((acc, it) => acc + it.received_quantity, 0);
}
export function sumCredited(items: ShipmentItem[]): number {
  return items.reduce((acc, it) => acc + it.credited_quantity, 0);
}
